File: ledger-models-java/src/main/java/common/util/LinkCache.java
package common.util;

import fintekkers.models.portfolio.PortfolioProto;
import fintekkers.models.price.PriceProto;
import fintekkers.models.security.SecurityProto;
import fintekkers.models.transaction.TransactionProto;

import java.time.Duration;
import java.time.Instant;
import java.time.ZonedDateTime;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Process-wide cache for proto bodies backing link-mode wrappers. Single
 * slot per UUID with asOf-validating reads, bounded LRU eviction, and
 * per-entity TTL bounds on null-asOf reads.
 *
 * <p><b>Read semantics</b> ({@link #get(UUID, ZonedDateTime)}):
 * <ul>
 *   <li>{@code requestedAsOf == null} ("latest acceptable") — cache hit
 *       allowed; subject to {@code ttlForLatest} to bound cross-process
 *       staleness this process can't observe (e.g. another process wrote
 *       a new version after we cached the prior one). Past TTL ⇒ miss.</li>
 *   <li>{@code requestedAsOf != null} (bitemporal-precise) — cache hit only
 *       when the cached entry's asOf equals the requested. No TTL — history
 *       doesn't change, so a past vintage cached arbitrarily long is fine.</li>
 * </ul>
 *
 * <p><b>Write semantics</b> ({@link #put(UUID, Object, ZonedDateTime)}):
 * newest-vintage wins. An older-vintage put does not evict a newer cached
 * entry — that preserves the cache's "newest known" invariant for
 * {@code requestedAsOf == null} reads.
 *
 * <p><b>Eviction:</b> bounded LRU. When {@code put} causes the map to
 * exceed its cap, the least-recently-used entry is removed. {@code get}
 * bumps recency.
 *
 * <p>Static singletons {@link #SECURITY}, {@link #PORTFOLIO}, {@link #PRICE},
 * {@link #TRANSACTION} are the process-wide caches for each entity type,
 * each constructed with TTL + cap matching its typical access pattern
 * (Portfolio + Security change slowly so TTL is 1 day; Price + Transaction
 * change quickly so TTL is short). Tests can create scoped instances via
 * the public constructor.
 *
 * <p>See {@code docs/adr/lazy-link-hydration.md} for the design rationale.
 */
public final class LinkCache<V> {

    /** Default TTL for null-asOf reads when no entity-specific value is provided. */
    public static final Duration DEFAULT_TTL_FOR_LATEST = Duration.ofSeconds(600);

    /** Default max entries when no entity-specific cap is provided. */
    public static final int DEFAULT_MAX_ENTRIES = 10_000;

    // Per-entity singletons. Tuned for the access pattern of each entity
    // type; production callers should use these and not the no-arg ctor.
    public static final LinkCache<SecurityProto>    SECURITY    =
            new LinkCache<>(Duration.ofDays(1), 100_000);
    public static final LinkCache<PortfolioProto>   PORTFOLIO   =
            new LinkCache<>(Duration.ofDays(1), 10_000);
    public static final LinkCache<PriceProto>       PRICE       =
            new LinkCache<>(Duration.ofSeconds(30), 200_000);
    public static final LinkCache<TransactionProto> TRANSACTION =
            new LinkCache<>(Duration.ofMinutes(1), 100_000);

    private final Duration ttlForLatest;
    private final int maxEntries;
    // accessOrder=true so .get() bumps recency; removeEldestEntry caps the
    // size automatically on .put(). Synchronized on this instance — simple
    // and correct; perf is good enough for the typical lazy-hydrate access
    // pattern (mostly reads, low contention).
    private final LinkedHashMap<UUID, CacheEntry<V>> map;

    public LinkCache() {
        this(DEFAULT_TTL_FOR_LATEST, DEFAULT_MAX_ENTRIES);
    }

    public LinkCache(Duration ttlForLatest) {
        this(ttlForLatest, DEFAULT_MAX_ENTRIES);
    }

    public LinkCache(Duration ttlForLatest, int maxEntries) {
        if (ttlForLatest == null || ttlForLatest.isNegative()) {
            throw new IllegalArgumentException(
                "ttlForLatest must be non-null and non-negative; got " + ttlForLatest);
        }
        if (maxEntries <= 0) {
            throw new IllegalArgumentException(
                "maxEntries must be > 0; got " + maxEntries);
        }
        this.ttlForLatest = ttlForLatest;
        this.maxEntries = maxEntries;
        this.map = new LinkedHashMap<>(16, 0.75f, /* accessOrder = */ true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<UUID, CacheEntry<V>> eldest) {
                return size() > LinkCache.this.maxEntries;
            }
        };
    }

    /**
     * @param id            the entity uuid
     * @param requestedAsOf null = "latest acceptable" (TTL-bounded);
     *                      non-null = exact-vintage match required (no TTL)
     * @return the cached value if the lookup is a hit per the read semantics
     *         above; otherwise null (caller must refetch)
     */
    public synchronized V get(UUID id, ZonedDateTime requestedAsOf) {
        CacheEntry<V> entry = map.get(id);
        if (entry == null) return null;
        if (requestedAsOf == null) {
            if (Duration.between(entry.cachedAt, Instant.now()).compareTo(ttlForLatest) > 0) {
                return null;
            }
            return entry.value;
        }
        if (entry.asOf == null) return null;
        return entry.asOf.equals(requestedAsOf) ? entry.value : null;
    }

    /**
     * Newest-wins write: if a cached entry for {@code id} already exists with
     * an asOf strictly after {@code asOf}, the incoming write is ignored
     * (but recency is still bumped — the caller saw the entry).
     */
    public synchronized void put(UUID id, V value, ZonedDateTime asOf) {
        if (id == null || value == null) {
            throw new IllegalArgumentException(
                "id / value must be non-null; got id=" + id +
                " value=" + (value == null ? "null" : "<non-null>"));
        }
        CacheEntry<V> existing = map.get(id);
        if (existing != null && existing.asOf != null && asOf != null && existing.asOf.isAfter(asOf)) {
            // Recency bump only — older vintage doesn't displace newer cached entry.
            return;
        }
        map.put(id, new CacheEntry<>(value, asOf, Instant.now()));
    }

    public synchronized void evict(UUID id) {
        map.remove(id);
    }

    public synchronized void clear() {
        map.clear();
    }

    public synchronized int size() {
        return map.size();
    }

    record CacheEntry<V>(V value, ZonedDateTime asOf, Instant cachedAt) {}
}

File: ledger-models-python/fintekkers/wrappers/util/link_cache.py
"""LinkCache — process-level cache of resolved entity protos keyed by uuid.

Companion to the Java common.util.LinkCache. Mirrors the same semantics:

* Single-slot per uuid (newest-asOf-wins on put).
* Reads validate the as_of: exact match returns the entry; mismatch returns
  None so the caller refetches the requested vintage.
* `as_of=None` on a read means "latest acceptable" — TTL-bounded; entries
  older than TTL_FOR_LATEST return None and force a refresh. Entries with
  a non-null as_of never TTL-expire (bitemporal history doesn't change).
* Bounded LRU eviction — when the cache fills, the least-recently-used
  entry is dropped on the next `put`. Per-entity defaults reflect the
  typical access pattern: Portfolio + Security entities change slowly so
  TTL is a day; Price + Transaction change quickly so TTL is short.

See `docs/adr/lazy-link-hydration.md` for the design.
"""

from __future__ import annotations

import threading
import time
from collections import OrderedDict
from datetime import datetime
from typing import Generic, Optional, TypeVar
from uuid import UUID

T = TypeVar("T")

# Backwards-compat alias used by tests written against the original module.
DEFAULT_TTL_FOR_LATEST_SECONDS = 60.0


class _Entry(Generic[T]):
    __slots__ = ("value", "as_of", "cached_at_monotonic")

    def __init__(self, value: T, as_of: Optional[datetime], cached_at_monotonic: float):
        self.value = value
        self.as_of = as_of
        self.cached_at_monotonic = cached_at_monotonic


class LinkCache(Generic[T]):
    """Process-wide cache, bounded LRU, per-entity TTL for null-asOf reads.

    Construct with `ttl_for_latest_seconds` for the null-asOf TTL and
    `max_entries` for the LRU cap. Defaults: 60s TTL, 10_000 entries.
    The four shipped singletons below override both.
    """

    def __init__(
        self,
        ttl_for_latest_seconds: float = DEFAULT_TTL_FOR_LATEST_SECONDS,
        max_entries: int = 10_000,
    ):
        if max_entries <= 0:
            raise ValueError(f"max_entries must be > 0, got {max_entries}")
        self._ttl = ttl_for_latest_seconds
        self._max_entries = max_entries
        # OrderedDict so we can move-to-end on access (LRU) and popitem(last=False)
        # the least-recently-used on overflow.
        self._map: "OrderedDict[UUID, _Entry[T]]" = OrderedDict()
        self._lock = threading.Lock()

    def get(self, uuid: UUID, requested_as_of: Optional[datetime]) -> Optional[T]:
        with self._lock:
            entry = self._map.get(uuid)
            if entry is None:
                return None
            if requested_as_of is None:
                if time.monotonic() - entry.cached_at_monotonic > self._ttl:
                    return None
            else:
                if entry.as_of is None:
                    return None
                if entry.as_of != requested_as_of:
                    return None
            # Hit — bump recency.
            self._map.move_to_end(uuid)
            return entry.value

    def put(self, uuid: UUID, value: T, as_of: Optional[datetime]) -> None:
        with self._lock:
            existing = self._map.get(uuid)
            if existing is not None and existing.as_of is not None and as_of is not None:
                if existing.as_of > as_of:
                    # Older-vintage write never displaces a newer cached entry.
                    # Still bump recency: the caller saw a fresh reference.
                    self._map.move_to_end(uuid)
                    return
            self._map[uuid] = _Entry(value, as_of, time.monotonic())
            self._map.move_to_end(uuid)
            while len(self._map) > self._max_entries:
                self._map.popitem(last=False)

    def evict(self, uuid: UUID) -> None:
        with self._lock:
            self._map.pop(uuid, None)

    def clear(self) -> None:
        with self._lock:
            self._map.clear()

    def size(self) -> int:
        with self._lock:
            return len(self._map)


# Per-entity TTL + cap defaults. Captured in the original LinkCache.java
# planning note. Tune freely — these are bounded heuristics, not invariants.
#
#   Portfolio / Security: 1-day TTL on null-as_of reads (entities change
#       infrequently); large caps because the universe is large but each
#       entry is small.
#   Transaction: 1-minute TTL (high churn).
#   Price: 30-second TTL (very high churn).
SECURITY: LinkCache = LinkCache(ttl_for_latest_seconds=86_400.0, max_entries=100_000)
PORTFOLIO: LinkCache = LinkCache(ttl_for_latest_seconds=86_400.0, max_entries=10_000)
PRICE: LinkCache = LinkCache(ttl_for_latest_seconds=30.0, max_entries=200_000)
TRANSACTION: LinkCache = LinkCache(ttl_for_latest_seconds=60.0, max_entries=100_000)
